### PR TITLE
Added Custom Code, Client, Category fields on coupon dashboard

### PR DESCRIPTION
**Pull request by iivic**
@mjfrey @marjev @vkaracic  I added Custom Code, Client, Category fields to the coupon dashboard and did some cleanup. We still have one search input which searches all the fields in the table.

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @iivic! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?number=974&repo=edx%2Fecommerce

**Comment by iivic** [comment]
screen shot 2016-10-19 at 14 15 31

File: ecommerce/extensions/api/serializers.py [hunk 1]
```diff
@@ -43,6 +43,55 @@
 PRODUCT_DETAIL_VIEW = 'api:v2:product-detail'
 
 
+def is_custom_code(obj):
+    """Helper method to check if the voucher contains custom code. """
+    if not is_enrollment_code(obj) and retrieve_quantity(obj) == 1:
```
Comment (marjev) [inline ecommerce/extensions/api/serializers.py]: simply `return not is_enrollment_code(obj) and retrieve_quantity(obj) == 1`

File: ecommerce/extensions/api/serializers.py [hunk 2]
```diff
@@ -43,6 +43,55 @@
...
+        return True
+    return False
+
+
+def is_enrollment_code(obj):
+    benefit = retrieve_voucher(obj).benefit
+    if benefit.type == Benefit.PERCENTAGE and benefit.value == 100:
```
Comment (marjev) [inline ecommerce/extensions/api/serializers.py]: `return benefit.type == Benefit.PERCENTAGE and benefit.value == 100`

File: ecommerce/extensions/api/serializers.py [hunk 3]
```diff
@@ -452,78 +516,57 @@ class CouponSerializer(ProductPaymentInfoMixin, serializers.ModelSerializer):
...
     def get_benefit_type(self, obj):
-        return self.retrieve_benefit(obj).type
+        return retrieve_benefit(obj).type
 
     def get_benefit_value(self, obj):
-        return self.retrieve_benefit(obj).value
+        return retrieve_benefit(obj).value
+
+    def get_catalog_query(self, obj):
+        offer = retrieve_offer(obj)
```
Comment (marjev) [inline ecommerce/extensions/api/serializers.py]: simply `return retrieve_offer(obj).condition.range.catalog_query`

**Review by marjev — COMMENTED** [review]
Other than a few comments, LGTM 👍

File: ecommerce/extensions/api/serializers.py [hunk 4]
```diff
@@ -43,6 +43,51 @@
...
+    benefit = retrieve_voucher(obj).benefit
+    return benefit.type == Benefit.PERCENTAGE and benefit.value == 100
+
+
+def retrieve_benefit(obj):
+    """Helper method to retrieve the benefit from voucher. """
+    return retrieve_voucher(obj).benefit
+
+
+def retrieve_end_date(obj):
```
Comment (vkaracic) [inline ecommerce/extensions/api/serializers.py:61]: Was it necessary to factor out all of these functions?

Comment (iivic) [reply ecommerce/extensions/api/serializers.py:61]: Some of them repeated in different serializers and they are not necessarily linked to one particular serializer. I concluded that they could be used in different serializers in the future  factoring them out will help to achieve that and prevent unnecessary repeating.

Comment (vkaracic) [reply ecommerce/extensions/api/serializers.py:61]: Alright.

**Review by vkaracic — APPROVED** [review]
Apart from the failing coverage, LGTM.

**Comment by iivic** [comment]
Review the last commit please.

File: ecommerce/extensions/api/v2/tests/views/test_coupons.py [hunk 5]
```diff
@@ -240,6 +240,24 @@ def setUp(self):
...
+            'client': 'TeštX',
+            'code': '123456',
+            'end_datetime': str(now() + datetime.timedelta(days=10)),
+            'price': 100,
+            'quantity': 1,
+            'start_datetime': str(now() - datetime.timedelta(days=10)),
+            'stock_record_ids': [1, 2],
+            'title': 'Tešt čoupon 2',
+            'voucher_type': Voucher.SINGLE_USE
+        }
```
Comment (marjev) [inline ecommerce/extensions/api/v2/tests/views/test_coupons.py]: Can we DRY this? It seems like you're only changing two values. The rest of them are same as `data`'s.

**Review by marjev — APPROVED** [review]

File: ecommerce/extensions/api/v2/tests/views/test_coupons.py [hunk 6]
```diff
@@ -309,7 +328,10 @@ def test_list_coupons(self):
         response = self.client.get(COUPONS_LINK)
         self.assertEqual(response.status_code, status.HTTP_200_OK)
         coupon_data = json.loads(response.content)['results'][0]
-        self.assertEqual(coupon_data['title'], self.data['title'])
+        self.assertEqual(coupon_data['title'], self.data2['title'])
+        self.assertEqual(coupon_data['category']['name'], self.data2['category']['name'])
+        self.assertEqual(coupon_data['client'], self.data2['client'])
+        self.assertEqual(coupon_data['code'], self.data2['code'])
```
Comment (vkaracic) [inline ecommerce/extensions/api/v2/tests/views/test_coupons.py]: Create a new test case `test_list_custom_code_coupons`, adjust the `self.data` values in that test case and test the response there. No need to juggle two different coupon data throughout the test class.

**Comment by iivic** [comment]
Comments addressed. Review latest commit.

**Comment by vkaracic** [comment]
LGTM

**Pull request merged**